src/Streamly/Internal/Data/Parse.hs

```diff
@@ -252,3 +277,101 @@ endBefore = undefined
+{-# ANN type ParState Fuse #-}
+data ParState sL sR a =
+    ParPair !([a], Either sL sL, [a], [a]) !([a], Either sR sR, [a], [a])
+
+-- XXX we can write a much simpler zipWith if the types do not have Back or
+-- Hold. Perhaps we can write a simpler partial function assuming that and
+-- statically assert the assumptions with the help of the compiler.
+--
+-- zipWithMin -- end when either stream ends
+-- zipWithFst -- end when the first stream ends
+--
+-- XXX requires -funfolding-use-threshold=200 to fuse
+--
+-- | @zipWith f p1 p2@ distributes its input to both @p1@ and @p2@ and combines
+-- their output using @f@.
+--
+-- /Internal/
+--
+{-# INLINE zipWith #-}
+zipWith :: Monad m
+    => (a -> b -> c) -> Parse m x a -> Parse m x b -> Parse m x c
+zipWith zf (Parse stepL initialL extractL) (Parse stepR initialR extractR) =
+    Parse step initial extract
+
+    where
+
+    {-# INLINE_LATE initial #-}
+    initial = do
+        sL <- initialL
+        sR <- initialR
+        return $ ParPair ([], Right sL, [], []) ([], Right sR, [], [])
+
+    {-# INLINE_LATE step #-}
+    step (ParPair (bufL, Right sL, inpL1, inpL2)
+                  (bufR, Right sR, inpR1, inpR2)) x = do
+        l <- useStream bufL inpL1 inpL2 stepL sL x
+        r <- useStream bufR inpR1 inpR2 stepR sR x
+        return $ Keep 0 $ ParPair l r
+
+    step (ParPair (bufL, Right sL, inpL1, inpL2) r@(_, Left _, _, _)) x = do
+        l <- useStream bufL inpL1 inpL2 stepL sL x
+        return $ Keep 0 $ ParPair l r
+
+    step (ParPair l@(_, Left _, _, _) (bufR, Right sR, inpR1, inpR2)) x = do
+        r <- useStream bufR inpR1 inpR2 stepR sR x
+        return $ Keep 0 $ ParPair l r
+
+    step s _ = return $ Keep 0 s
+
+    -- consume one input item and return the next state of the fold
+    {-# INLINE useStream #-}
+    useStream buf inp1 inp2 stp st y = do
+        let (x, inp11, inp21) =
+                case inp1 of
+                    [] -> (y, [], [])
+                    z : [] -> (z, reverse (x:inp2), [])
+                    z : zs -> (z, zs, x:inp2)
+        r <- stp st x
+        let buf1 = x:buf
+        return $ case r of
+            Hold s -> (buf1, Right s, inp11, inp21)
+            Keep n s ->
+                 assert (n <= length buf1)
+                        (Prelude.take n buf1, Right s, inp11, inp21)
+            Back n s ->
+                let (src0, buf2) = splitAt n buf1
+                    src  = Prelude.reverse src0
+                 in assert (n <= length buf1)
+                           (buf2, Right s, src ++ inp11, inp21)
+            Halt s -> (buf1, Left s, inp11, inp21)
+
+    {-# INLINE extractBoth #-}
+    extractBoth sL sR = do
+        rL <- extractL sL
+        case rL of
+            Left err -> return $ Left err
+            Right (_, x) -> do
+                rR <- extractR sR
+                return $ case rR of
+                    Left err -> Left err
+                    Right (_, y) -> Right (0, zf x y)
```

Comment:
Is `Right (0 ...` correct?
Shouldn't it be more like `Right (min (length bufL1 + length bufL2) (length bufR1 + length bufR2)) ...`